add support for Pico Keyboard

qufb opened the pull request:
Moved from https://github.com/libretro/picodrive/pull/205

---

These changes introduce a new input method for the Pico Keyboard, a PS/2 peripheral for the Sega Pico, allowing emulation of games that use this peripheral:

* Keyboard Pico - Kantan Waku Waku Keyboard
* Keyboard Pico 2 - Sawattemiyou! Yoiko no Hajimete Keyboard
* Kitty to Minna no Keyboard Pico - Kitty to Minna no Hajimete no Keyboard!
* Kibodeu Piko

Requires a dedicated input update function for frontends, implemented in my libpicofe pull request (https://github.com/irixxxx/libpicofe/pull/1). Once that one is merged, I can update the submodule repository in this one.

Since the original input handling used a bitmask for mapping keys to events, the larger number of keys required for this peripheral had to be implemented in another way, in this case it uses an array that is associated with a single input binding type (IN_BINDTYPE_PICO_PS2).

In order to sufficiently test the keyboard, I found that activating most minigames requires interacting with the Storyware via the pen button. Since these games don’t render a pen cursor, I added a pen position overlay to guide a player, disabled by default.

Tested under Linux with the generic SDL frontend.

Limitations

* Key bindings are hardcoded at compile time (see ./platform/common/input_pico.h and ./platform/common/inputmap_kbd.h), missing logic to allow reading and storing to configuration file.

For more context, see the following documentation: https://qufb.gitlab.io/writeups/pico#keyboard-pico

irixxxx commented:
sorry it takes so long. I'm currently busy with a lot of offline tasks which prevents me from doing much development.
I'm looking into a more general approach, to include the SC3000 as well.

qufb commented:
No worries. I also haven't look much into the libretro part, but it will probably be similar to the dedicated keyboard input that was done some time ago for a Dreamcast core (https://www.libretro.com/index.php/reicast-libretro-now-supports-the-keyboard-as-an-input-peripheral/)

irixxxx commented:
@qufb are you still available for this? I have finally come around to do the preparations for SC-3000 support, hence we could now discuss how to generalize your approach in libpicofe and picodrive in a way that would allow using it for both pico and SC-3000.

qufb commented:
Right now I'm focused on other projects, but you can leave some comments on the pull request and I'll try to address them when I have time.

From what I recall, the new input update functions on libpicofe are probably what is more convenient to generalize for any keyboard-based input. The rest of the stuff like key mappings will be different.

irixxxx commented:
This seems to be getting fairly complicated when considering international keyboard layouts:
- the SDL1.2 keysym contains sym and mod. sym has the base key value, no matter what the modifier state is, so e.g. pressing the 'a' key alwaya gives you 'a' as sym, no matter what modifiers are additionally pressed. While this is rather straight forward for alpha characters, it is by no means for anything else. You can't even count on the number keys, if you look at the french layout those are only accessible with the shift modifier. I would need to know the current layout, which SDL isn't providing me with.
- the scancode is hardware dependent and as such unusable, so just mapping keys 1:1 to the desired layout ignoring what's the actual mapping is in the users' layout isn't working.
- the only way seems to be to get the Unicode value for the character, but that would require mapping to the desired keyboard matrix, including simulating modifier keys if needed.

International keyboard layout can be quite different even for Latin letter keyboards, just look at the french one. What a mess.

irixxxx commented:
I've finally come back to this. I've rebased your changes onto the current state of affairs, and if you're still there, I'd encourage you to have a look at my kbd branch for checking, since I had to change some things to merge this with my latest.

BTW, Pico support has improved greatly. You might want to have a look.

irixxxx commented:
added some basic code for keyboard configuration.

irixxxx commented:
@qufb I did some cleanup, added support for configuring the key mapping, and also added basic support for SC-3000.

If by any chance you are still interested in any of this, have a look and give me some feedback please. I'm going to close this PR after adding the kbd branch, which contains your changes as well as mine, to master.

irixxxx commented:
First, thanks for looking after all this time. Sorry for this. I had this version 2.00 with all the things I wanted into it looming before me for much longer than I anticipated.

I'll fix the page addressing in menu.

Yes, it's most probably not difficult to detect Kibodeu and maintain an entry in carthw.cfg, or even generally detect Korean games. The default mappings are however only valid for American keyboards. The good question is where most of the users for this potentially are. If most of them have to reconfigure the keyboard mapping anyway, would it be necessary to maintain different defaults? Or would it maybe confuse the users more that the keyboard behaves differently on some (well, one?) games?

irixxxx commented:
I have changed some small things, like texts in the mapping page, adding a kbd switch to turn off all emulator hotkeys except the keyboard switch, page 7, added unbinding, some bugs fixed. Apart from that Korean mapping problem, does it look OK?

irixxxx commented:
It's not completely done yet:
- get_ticks inside the emulation core should move to platform directory
- there's no virtual keyboard
- no libretro version
but I think development can continue in master, now that it compiles and basically works on both Pico and SC-3000

qufb commented:
@irixxxx I was only able to check these now.

> I have changed some small things, like texts in the mapping page, adding a kbd switch to turn off all emulator hotkeys except the keyboard switch, page 7, added unbinding, some bugs fixed. Apart from that Korean mapping problem, does it look OK?

Looks good, maybe it makes sense to also enable "MD/Genesis/Pico options > Pico Keyboard" when pressing the F12 keybind to enable the keyboard, as it can be "on" but not actually enabled.

> Yes, it's most probably not difficult to detect Kibodeu and maintain an entry in carthw.cfg, or even generally detect Korean games. The default mappings are however only valid for American keyboards. The good question is where most of the users for this potentially are. If most of them have to reconfigure the keyboard mapping anyway, would it be necessary to maintain different defaults? Or would it maybe confuse the users more that the keyboard behaves differently on some (well, one?) games?

I think it's best for the keyboard configuration screen layout to show a consistent layout of the target emulated keyboard. For the Hangul layout, it would be nice to show the brackets side-by-side, as it is on real hardware.

For now, I guess you can assume an English speaking audience as the majority, and only have defaults for the US ANSI 104 keyboard. Having config templates for other layouts seems like the most straightforward way to iterate on this, they can always be added on demand.

irixxxx commented:
> Looks good, maybe it makes sense to also enable "MD/Genesis/Pico options > Pico Keyboard" when pressing the F12 keybind to enable the keyboard, as it can be "on" but not actually enabled.

I avoided changing things in the configuration automatically up to now. I believe this might confuse the user more than it's helpful. How about displaying "keyboard disabled in options" or something like this when hitting the key while kbd isn't enabled in the config?

> > Yes, it's most probably not difficult to detect Kibodeu and maintain an entry in carthw.cfg, or even generally detect Korean games. The default mappings are however only valid for American keyboards. The good question is where most of the users for this potentially are. If most of them have to reconfigure the keyboard mapping anyway, would it be necessary to maintain different defaults? Or would it maybe confuse the users more that the keyboard behaves differently on some (well, one?) games?
> 
> I think it's best for the keyboard configuration screen to show a consistent layout of the target emulated keyboard. For the Hangul layout, it would be nice to show the brackets side-by-side, as it is on real hardware.

The keyboard configuration is saved in the config files, and besides, you can save an auto-loaded config file per game. I admit the user must do something by themselves, but OTOH, there *is* a possibility to go with, and it needs to be done only once.

> For now, I guess you can assume an English speaking audience as the majority, and only have defaults for the US ANSI 104 keyboard. Having config templates for other layouts seems like the most straightforward way to iterate on this, they can always be added on demand.

I hope so. But the same reasoning as before holds: The user can configure the keys to his liking and save the global configuration. After that, ideally they shouldn't rarely to never need to adjust this again.

qufb commented:
> I avoided changing things in the configuration automatically up to now. I believe this might confuse the user more than it's helpful. How about displaying "keyboard disabled in options" or something like this when hitting the key while kbd isn't enabled in the config?

Sounds good, makes it clear what to do.

irixxxx commented:
virtual keyboard added, selectable in the controls configuration page.
